### PR TITLE
feat: support 'bin' function push down

### LINKED ISSUE #5103: Call For Participation: implement `Bin` function push down
## Feature Request

**Describe the feature you'd like:**

We'd like to support `Bin` function push down.

**Teachability, Documentation, Adoption, Migration Strategy:**

At least one integration test and several unit tests.

### PR DESCRIPTION
**Pull request by An-DJ**
### What problem does this PR solve?

Issue Number: close #5103 

Problem Summary: Implement Bin function push down

### What is changed and how it works?
Support Bin function push down.

### Check List

Tests 

- [X] Unit test
- [X] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: tests/fullstack-test/expr/bin.test [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+# limitations under the License.
+
+mysql> drop table if exists test.t1;
+mysql> create table test.t1(c1 bigint);
+mysql> insert into test.t1 values(0);
+mysql> insert into test.t1 values(1);
+mysql> insert into test.t1 values(44);
+mysql> insert into test.t1 values(100);
+mysql> insert into test.t1 values(-9223372036854775808);
+mysql> insert into test.t1 values(9223372036854775807);
```
Comment (bestwoody) [inline tests/fullstack-test/expr/bin.test:22]: better to add case of `null` value

File: dbms/src/Functions/FunctionsString.cpp [hunk 2]
```diff
@@ -4910,6 +4911,122 @@ class FunctionFormatWithLocale : public IFunction
...
+        if (!first_argument->isInteger())
+            throw Exception(
+                fmt::format("Illegal type {} of first argument of function {}", first_argument->getName(), getName()),
+                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+        return std::make_shared<DataTypeString>();
+    }
+
+    template <typename T>
+    static void executeOneUIntOrInt(T xx, char *& out, bool skip_leading_zero = true, bool auto_close = true)
```
Comment (bestwoody) [inline dbms/src/Functions/FunctionsString.cpp:4946]: please use a meaningful name for 1st param.

File: dbms/src/Functions/FunctionsString.cpp [hunk 3]
```diff
@@ -4910,6 +4911,122 @@ class FunctionFormatWithLocale : public IFunction
...
+        {
+            return false;
+        }
+    }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result) const override
+    {
+        const IColumn * column = block.getByPosition(arguments[0]).column.get();
+        ColumnPtr res_column;
+        if (tryExecuteUIntOrInt<UInt8>(column, res_column) || tryExecuteUIntOrInt<UInt16>(column, res_column) || tryExecuteUIntOrInt<UInt32>(column, res_column) || tryExecuteUIntOrInt<UInt64>(column, res_column) || tryExecuteUIntOrInt<Int8>(column, res_column) || tryExecuteUIntOrInt<Int16>(column, res_column) || tryExecuteUIntOrInt<Int32>(column, res_column) || tryExecuteUIntOrInt<Int64>(column, res_column))
```
[the remark this answers is not in the document]
Comment (An-DJ) [reply dbms/src/Functions/FunctionsString.cpp:5013]: @bestwoody 
I try to split that code to multiple lines but it cannot pass the `Format Check`. More details:
https://ci.pingcap.net/blue/organizations/jenkins/tiflash-build-common/detail/tiflash-build-common/8189/pipeline/

After format it with `clang-format` or `format.diff.py`, the multiple lines is formatted into a single line like below.

Maybe the format rules should be changed : (
```
diff --git a/dbms/src/Functions/FunctionsString.cpp b/dbms/src/Functions/FunctionsString.cpp
index 8c2d25b78..0dcb618b1 100644
--- a/dbms/src/Functions/FunctionsString.cpp
+++ b/dbms/src/Functions/FunctionsString.cpp
@@ -5010,14 +5010,7 @@ public:
     {
         const IColumn * column = block.getByPosition(arguments[0]).column.get();
         ColumnPtr res_column;
-        if (tryExecuteUIntOrInt<UInt8>(column, res_column) ||
-           tryExecuteUIntOrInt<UInt16>(column, res_column) ||
-           tryExecuteUIntOrInt<UInt32>(column, res_column) ||
-           tryExecuteUIntOrInt<UInt64>(column, res_column) ||
-           tryExecuteUIntOrInt<Int8>(column, res_column) ||
-           tryExecuteUIntOrInt<Int16>(column, res_column) ||
-           tryExecuteUIntOrInt<Int32>(column, res_column) ||
-           tryExecuteUIntOrInt<Int64>(column, res_column))
+        if (tryExecuteUIntOrInt<UInt8>(column, res_column) || tryExecuteUIntOrInt<UInt16>(column, res_column) || tryExecuteUIntOrInt<UInt32>(column, res_column) || tryExecuteUIntOrInt<UInt64>(column, res_column) || tryExecuteUIntOrInt<Int8>(column, res_column) || tryExecuteUIntOrInt<Int16>(column, res_column) || tryExecuteUIntOrInt<Int32>(column, res_column) || tryExecuteUIntOrInt<Int64>(column, res_column))
         {
             block.getByPosition(result).column = std::move(res_column);
             return;
```

File: dbms/src/Functions/tests/gtest_bin.cpp [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        executeFunction("bin", createColumn<Int64>({100})));
+}
+CATCH
+
+TEST_F(TestBin, Boundary)
+try
+{
+    ASSERT_COLUMN_EQ(
+        createColumn<String>({"1111111111111111111111111111111111111111111111111111111110000000", "1111111"}),
+        executeFunction("bin", createColumn<Int8>({INT8_MIN, INT8_MAX})));
```
[the remark this answers is not in the document]
Comment (An-DJ) [reply dbms/src/Functions/tests/gtest_bin.cpp]: Done

File: dbms/src/Functions/FunctionsString.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (bestwoody) [reply dbms/src/Functions/FunctionsString.cpp:5013]: ok

**Comment by bestwoody** [comment]
LGTM，please resolve conflicts at last。

**Review by LittleFall — APPROVED** [review]
lgtm, thanks!

**Pull request merged**